src/utils/actionToUrl.js

```diff
@@ -3,40 +3,40 @@ import { compileUrl, cleanBasename } from './index'
 
 import type {
   Route,
-  Payload,
-  Params,
-  RoutesMap,
+  Routes,
   ReceivedAction as Action,
   Options
 } from '../flow-types'
 
 export default (
   action: Action,
   api: Object,
-  prevRoute?: Object
-): string => {
-  const { routes, options: opts } = api
-  const { type, params, query, state, hash, basename } = action
+  prevRoute?: string
+): Object => {
+  const { routes, options: opts }: Object = api
+  const { type, params, query, state, hash, basename }: Action = action
 
   const route = routes[type] || {}
-  const path = typeof route === 'object' ? route.path : route
+  const path: string | void | string | void = typeof route === 'object' ? route.path : route
 
-  const p = formatParams(params, route, opts)
+  const p: void | {} = formatParams(params, route, opts)
   const q = formatQuery(query, route, opts)
-  const s = formatState(state, route, opts)
-  const h = formatHash(hash, route, opts)
+  const s: ?Object = formatState(state, route, opts)
+  const h: string = formatHash(hash, route, opts)
 
   const bn = cleanBasename(basename)
   const isWrongBasename = bn && !opts.basenames.includes(bn)
+  // $FlowFixMe
+  // TODO: Find out what needs to happen here.. ?not cool kyle?
   if (basename === '') s._emptyBn = true // not cool kyle
 
   try {
     if (isWrongBasename) {
       throw new Error(`[rudy] basename "${bn}" not in options.basenames`)
     }
 
-    const pathname = compileUrl(path, p, q, h, route, opts) || '/' // path-to-regexp throws for failed compilations; we made our queries + hashes also throw to conform
-    const url = bn + pathname
+    const pathname: string = compileUrl(path, p, q, h, route, opts) || '/' // path-to-regexp throws for failed compilations; we made our queries + hashes also throw to conform
+    const url: string = bn + pathname
```

Comment:
url's `string` type should be inferred if everything prior was done correctly.

the goal is to tightly type the interfaces (function args + returns) and let the flow compiler infer function bodies.

Reply:
So you mean control it higher up where the function is rather than lower down where a variable gets defined?